docs/internal/types.md

```diff
@@ -382,3 +432,44 @@ x = ("hi", 42); // Type inference looks at `x`'s type for literals
 // Access elements
 (var s, var i) = x; // Is this a destructive read? Great question!
 ```
+
+## Self type
+
+The `Self` type is used on interfaces in cases where the concrete type will be derived later.
+
+When concrete types implement an interface, they must implement all methods that the interface provides.
+However, the interface type can be implemented by any number of concrete classes, and so cannot know what would be the concrete type implementing it.
+That's where `Self` types come in: they're the concrete type implementing the interface.
+
+Example:
+```ts
+interface EqualComparable
+{
+  ==(Self & readonly, Self & readonly) : Bool;
+}
+
+interface Addable
+{
+  +(Self & readonly, Self & readonly) : Self & imm;
+}
+
+class MyClass : EqualComparable, Addable
+{
+  // Some internal field
+  field : U32 & mut;
+
+  // Implements the EqualComparable interface
+  ==(lhs : MyClass & readonly, rhs : MyClass & readonly) : Bool
+  {
+    lhs.field == rhs.field
+  }
+
+  // Implements the Addable interface
+  +lhs : MyClass & readonly, rhs : MyClass & readonly) : MyClass & imm
+  {
+    lhs.field + rhs.field
+  }
+}
+```
+
+In the example above, the `MyClass` argument type _matches_ the `Self` type in the interface declaration because it is the same type as the class declaring it.
```

Comment:
I think it would be useful to show why this is a problem without `Self`.  For example, if you defined:

```Verona
interface EqualComparable
{
  ==(EqualComparable & readonly, EqualComparable & readonly) : Bool;
}
```

Most classes would not conform to it because they can't be compared against *anything* that is equallity-comparable.  You'd need to define something like this:

```Verona
interface EqualComparable[T]
{
  ==(T & readonly, T & readonly) : Bool;
}
```

Except that this also isn't right: a concrete instance of the interface would define a single type that could be compared against itself and you want the interface to be generic over all concrete instances (`MyClass` would implement `EqualityComparable[MyClass]`, not `EqualityComparable` and if you have to specify the concrete type that implements the interface whenever you use the interface then there's no point in providing the interface).

Reply:
The first case obviously doesn't work, as you say.

The second case, as I understand, is not that _"there is no point"_, but that you'd have to use the interface as a generic, too, and you wouldn't be able to implement the method as the interface.

```ts
function(Interface[T]) : Bool {
  // I can't make it generic here because T has to be concrete of *any* type
}
...
class Foo : Interface[Foo] { ... }
class Bar : Interface[Bar] { ... }
foo = Foo...;
bar = Bar...;
...
function(foo) // matches function(Interface[Foo]);
function(bar) // matches function(Interface[Bar]);
```

Makes sense?

Reply:
Yes, you can't use the interface in its non-generic form in the second case, so it doesn't let you do the thing.